File: 프로그래머스/2/181187. 두 원 사이의 정수 쌍/두 원 사이의 정수 쌍.js
function solution(r1, r2) {
    let result = 0
    
    for (let x = 1; x <= r2; x++) { // x = 0 제외한 이유는 *4 할 때 개수가 중복되지 않게 하기 위함
        let maxY = Math.floor(Math.sqrt(r2*r2 - x*x)) 
        // x = 1일 때 : 9 - 1 = √8 = 2.xx = 2 
        // x = 2일 때 : 9 - 4 = √5 = 2.xx = 2
        // x = 3일 때 : 9 - 9 = 0
        let minY = r1 <= x ? 0 : Math.ceil(Math.sqrt(r1*r1 - x*x)) 
        // x = 1일 때 : 4 - 1 = √3 = 1.xx = 2
        // x = 2일 때 : 4 - 4 = √0 = 0
        // x = 3일 때 : 0
        // console.log('x = ', x, '일 때 : y = ', maxY, ' ~ ', minY)
        
        result += maxY - minY + 1
    }
    
    return result * 4
}

// 시간초과
// function solution(r1, r2) {
//     let result = 0;
//     for (let x = -r2; x <= r2; x++) {
//         for (let y = -r2; y <= r2; y++) {
//             if (Math.sqrt(Math.pow(x, 2) + Math.pow(y, 2)) >= r1 && Math.sqrt(Math.pow(x, 2) + Math.pow(y, 2)) <= r2) {
//                 result++;
//             }
//         }
//     }
//     return result;
// }

File: 프로그래머스/2/181187. 두 원 사이의 정수 쌍/두 원 사이의 정수 쌍.py
# 원의 방정식
# x^2 + y^2 <= r^2

# r2 원 안에 있는 점의 개수 - r1 원 안에 있는 점의 개수
# 1사분면만 구해서 * 4
import math

def solution(r1, r2):
    answer = 0
    
    for x in range(0, r2 + 1):
        maxY = math.floor(math.sqrt(r2*r2 - x*x))
        minY = 0 if r1 < x else math.ceil(math.sqrt(r1*r1 - x*x)) 
        
        answer += maxY - minY + 1
        
    # 경계선 부분 중복되므로 빼주기
    answer -= r2 - r1 + 1
    
    return answer * 4